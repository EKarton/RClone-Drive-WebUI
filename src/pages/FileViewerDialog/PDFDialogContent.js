import { pdfjs, Document, Page } from 'react-pdf';
import { useState } from 'react';

pdfjs.GlobalWorkerOptions.workerSrc = `//cdnjs.cloudflare.com/ajax/libs/pdf.js/${pdfjs.version}/pdf.worker.min.js`;

export default function PDFDialogContent({ fileUrl }) {
  const [hasLoaded, setHasLoaded] = useState(false);
  const [numPages, setNumPages] = useState(null);

  const handlePdfLoadSuccess = ({ numPages }) => {
    setNumPages(numPages);
    setHasLoaded(true);
  };

  const renderAllPages = () => {
    return Array.from(new Array(numPages), (el, index) => (
      <Page key={`page_${index + 1}`} pageNumber={index + 1} />
    ));
  };

  return (
    <>
      <Document
        file={fileUrl}
        options={{ workerSrc: '/pdf.worker.js' }}
        onLoadSuccess={handlePdfLoadSuccess}
      >
        {hasLoaded && renderAllPages()}
      </Document>
    </>
  );
}
