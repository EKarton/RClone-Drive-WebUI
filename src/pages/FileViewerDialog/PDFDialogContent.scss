/* stylelint-disable */
.react-pdf__Page__canvas {
  margin: 0 auto;
  width: 100% !important;
  height: auto !important;
}
/* stylelint-enable */
