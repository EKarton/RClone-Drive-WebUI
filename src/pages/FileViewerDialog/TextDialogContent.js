import { DialogContent, DialogContentText } from '@mui/material';
import { useState, useEffect } from 'react';
import './TextDialogContent.scss';

export default function TextDialogContent({ fileBlob }) {
  const [data, setData] = useState(null);

  useEffect(() => {
    fileBlob.text().then((contents) => setData(contents));
  }, [fileBlob]);

  return (
    <DialogContent>
      <DialogContentText className="text-dialog-content">{data}</DialogContentText>
    </DialogContent>
  );
}
