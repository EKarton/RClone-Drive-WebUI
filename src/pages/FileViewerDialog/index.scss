.imageviewer-dialog {
  &__paper {
    max-height: 100%;
    max-width: 100%;
  }

  &__header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
  }

  &__header-content {
    color: white;
  }
}
