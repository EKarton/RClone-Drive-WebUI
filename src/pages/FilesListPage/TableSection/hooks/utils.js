import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
