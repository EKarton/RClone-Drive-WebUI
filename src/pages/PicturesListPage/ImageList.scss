.imagelist {
  &__row {
    display: flex;
    padding-bottom: 0.25rem;
  }
}
