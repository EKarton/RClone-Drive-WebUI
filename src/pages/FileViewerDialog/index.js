import { Dialog, DialogContent, IconButton } from '@mui/material';
import CircularProgress from '@mui/material/CircularProgress';
import useRCloneClient from 'hooks/useRCloneClient';
import { useEffect, useState } from 'react';
import { ImageMimeTypes } from 'utils/constants';
import PDFDialogContent from './PDFDialogContent';
import FileDownloadIcon from '@mui/icons-material/FileDownload';
import './index.scss';
import FileSaver from 'file-saver';
import useFileViewer from 'hooks/useFileViewer';

export default function FileViewerDialog() {
  const { fileInfo, isOpen, hide } = useFileViewer();
  const rCloneClient = useRCloneClient();

  const [error, setError] = useState();
  const [fileMimeType, setFileMimeType] = useState();
  const [fileBlob, setFileBlob] = useState();
  const [fileUrl, setFileUrl] = useState();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setFileUrl(undefined);
        setError(undefined);

        const { remote, folderPath, fileName } = fileInfo;
        const response = await rCloneClient.fetchFileContents(
          remote,
          folderPath,
          fileName
        );

        const mimeType = response.headers['content-type'];
        const blob = new Blob([response.data], { type: mimeType });

        setFileMimeType(mimeType);
        setFileBlob(blob);
        setFileUrl(URL.createObjectURL(blob));
      } catch (err) {
        setError(err);
      }
    };

    if (fileInfo?.remote && fileInfo?.folderPath && fileInfo?.fileName) {
      fetchData();
    }
  }, [fileInfo, rCloneClient]);

  const handleDownloadButtonClicked = () => {
    FileSaver.saveAs(fileBlob, fileInfo?.fileName);
  };

  const renderDownloadButton = () => {
    return (
      <div className="imageviewer-dialog__header">
        <div className="imageviewer-dialog__header-content">{fileInfo?.fileName}</div>
        <div>
          <IconButton onClick={handleDownloadButtonClicked}>
            <FileDownloadIcon
              className="imageviewer-dialog__header-content"
              data-testid="download-button"
            />
          </IconButton>
        </div>
      </div>
    );
  };

  const renderDialogContent = () => {
    if (!fileUrl && !error) {
      return (
        <DialogContent>
          <CircularProgress />
        </DialogContent>
      );
    }

    if (error) {
      return <div data-testid="error-message">Error!</div>;
    }

    if (ImageMimeTypes.has(fileMimeType)) {
      return <img src={fileUrl} alt={fileInfo?.fileName} data-testid="image-content" />;
    }

    if (fileMimeType === 'application/pdf') {
      return <PDFDialogContent fileUrl={fileUrl} />;
    }

    return null;
  };

  return (
    <Dialog
      className="imageviewer-dialog"
      open={isOpen}
      onClose={hide}
      maxWidth="sm"
      classes={{ paper: 'imageviewer-dialog__paper' }}
    >
      {renderDownloadButton()}
      {renderDialogContent()}
    </Dialog>
  );
}
