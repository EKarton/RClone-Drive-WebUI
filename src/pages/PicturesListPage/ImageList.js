import LazyImageList from 'components/LazyImageList';
import useFileViewerDialog from 'hooks/utils/useFileViewerDialog';
import { StatusTypes } from 'utils/constants';
import LazyImageListSkeleton from 'components/LazyImageListSkeleton';
import useRecentlyViewedImages from 'hooks/utils/useRecentlyViewedImages';
import useFetchPictures from 'hooks/fetch-data/useFetchPictures';

export default function ImageList({ remote, path }) {
  const fileViewer = useFileViewerDialog();
  const recentlyViewedImages = useRecentlyViewedImages();
  const { status, data } = useFetchPictures(remote, path);

  if (status === StatusTypes.LOADING) {
    return <LazyImageListSkeleton />;
  }

  if (status === StatusTypes.ERROR) {
    return <div>Error!</div>;
  }

  const handleImageClicked = (image) => {
    recentlyViewedImages.addImage(image);
    fileViewer.show(image);
  };

  return (
    <LazyImageList images={data} remote={remote} onImageClicked={handleImageClicked} />
  );
}
