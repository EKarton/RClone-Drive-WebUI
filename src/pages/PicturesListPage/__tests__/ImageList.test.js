import LazyImageList from 'components/LazyImageList';
import { FileViewerDialogProvider } from 'contexts/FileViewerDialog/index';
import useFetchPictures from 'hooks/fetch-data/useFetchPictures';
import useFileViewerDialog from 'hooks/utils/useFileViewerDialog';
import { StatusTypes } from 'utils/constants';
import { hashRemotePath } from 'utils/remote-paths-url';
import { mockPictures } from 'test-utils/mock-responses';
import { act, customRender, waitFor } from 'test-utils/react';
import ImageList from '../ImageList';

jest.mock('hooks/fetch-data/useFetchPictures');
jest.mock('hooks/utils/useFileViewerDialog');
jest.mock('components/LazyImageList');

describe('ImageList', () => {
  const remote = 'googledrive';
  const fetchPicturesFn = jest.fn();
  const fileViewerShowFn = jest.fn();

  beforeEach(() => {
    useFetchPictures.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: mockPictures.list,
    });

    useFileViewerDialog.mockReturnValue({
      show: fileViewerShowFn,
    });

    LazyImageList.mockReturnValue(null);
  });

  it('should match snapshots when the api call succeeds', async () => {
    const { baseElement } = renderComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it('should render skeleton when api is being loaded', async () => {
    useFetchPictures.mockReturnValue({
      status: StatusTypes.LOADING,
    });

    const { baseElement } = renderComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it('should render error page when fetching data was not successful', async () => {
    useFetchPictures.mockReturnValue({
      status: StatusTypes.ERROR,
      error: new Error('Error!'),
    });

    expect(renderComponent).toThrowError();
  });

  it('should call fileViewer.show() when someone clicks on an image', async () => {
    jest.useFakeTimers();

    fetchPicturesFn.mockResolvedValue(mockPictures.list);

    const fileInfo = { remote, dirPath: 'Pictures', fileName: 'image.png' };

    LazyImageList.mockImplementation(({ onImageClicked }) => {
      setTimeout(() => onImageClicked(fileInfo), 10000);
      return null;
    });

    renderComponent();

    act(() => jest.runAllTimers());

    await waitFor(() => expect(fileViewerShowFn).toBeCalledWith(fileInfo));
  });

  const renderComponent = () => {
    const route = `/pictures/${hashRemotePath(`${remote}:`)}`;

    const initialRCloneInfoState = {
      endpoint: 'http://localhost:5572',
      username: 'local',
      password: '1234',
    };

    const component = (
      <FileViewerDialogProvider>
        <ImageList remote={remote} path="" />
      </FileViewerDialogProvider>
    );

    return customRender(component, { initialRCloneInfoState }, { route });
  };
});
