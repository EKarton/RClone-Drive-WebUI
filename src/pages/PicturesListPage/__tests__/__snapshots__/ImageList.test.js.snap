// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageList should match snapshots when the api call succeeds 1`] = `
<body>
  <div />
</body>
`;

exports[`ImageList should render skeleton when api is being loaded 1`] = `
<body>
  <div>
    <div
      style="overflow: visible; height: 0px; width: 0px;"
    />
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1px; height: 1px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </div>
</body>
`;
